flink-architecture-tests/src/test/java/org/apache/flink/architecture/rules/ApiAnnotationRules.java

```diff
@@ -0,0 +1,168 @@
+                            .notBeAnnotatedWith(PublicEvolving.class)
+                            .andShould()
+                            .notBeAnnotatedWith(Public.class)
+                            .as(
+                                    "API visibility annotations must be declared on top-level (not nested) classes."));
+
+    /**
+     * This is a stronger requirement than {@link
+     * #PUBLIC_EVOLVING_API_METHODS_USE_ONLY_PUBLIC_EVOLVING_API_TYPES} and ensures that {@link
+     * Public} APIs do not use {@link PublicEvolving} APIs.
+     */
+    @ArchTest
+    public static final ArchRule PUBLIC_API_METHODS_USE_ONLY_PUBLIC_API_TYPES =
+            freeze(
+                    methods()
+                            .that()
+                            .areDeclaredInClassesThat(
+                                    areJavaClasses().and(annotatedWith(Public.class)))
+                            .and()
+                            .arePublic()
+                            .and()
+                            .areNotAnnotatedWith(Internal.class)
+                            .should(
+                                    haveLeafReturnTypes(
+                                            resideOutsideOfPackage("org.apache.flink..")
+                                                    .or(
+                                                            areAnnotatedWithAtLeastOneOf(
+                                                                    Public.class,
+                                                                    Deprecated.class))))
+                            .andShould(
+                                    haveLeafArgumentTypes(
+                                            resideOutsideOfPackage("org.apache.flink..")
+                                                    .or(
+                                                            areAnnotatedWithAtLeastOneOf(
+                                                                    Public.class,
+                                                                    Deprecated.class))))
+                            .as(
+                                    "Return and argument types of methods annotated with @Public must be annotated with @Public, too."));
+
+    @ArchTest
+    public static final ArchRule PUBLIC_EVOLVING_API_METHODS_USE_ONLY_PUBLIC_EVOLVING_API_TYPES =
+            freeze(
+                    methods()
+                            .that()
+                            .areDeclaredInClassesThat(
+                                    areJavaClasses()
+                                            .and(
+                                                    areAnnotatedWithAtLeastOneOf(
+                                                            Public.class, PublicEvolving.class)))
+                            .and()
+                            .arePublic()
+                            .and()
+                            .areNotAnnotatedWith(Internal.class)
```

Comment:
Yes, that is a fair point. I haven't dug into all violations in detail yet, but I can check whether this changes the identified violations. If not, then there wouldn't be a reason to add this, since this scenario wouldn't be able to arise anymore.